### ISSUE TITLE
STEP: export loses colors for certain Bodies

### ISSUE THREAD

**Issue by mwoehlke** [FreeCAD/FreeCAD#10876]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

STEP export loses colors for certain assemblages.

To reproduce:
1. Create a Body and change its color
2. Add an additive cube primitive with default settings
3. Create a LinearPattern feature with the preceding (change length to 10mm)
4. Add an additive cylinder primitive with default settings

You should end up with something vaguely resembling a simple pea whistle, although any shape that forms a valid body using the above combination of features should work. The exact shape doesn't matter; the problem seems to manifest when adding features to a LinearPattern. (I didn't check other pattern types.) Naturally, the model on which I discovered this is quite a bit more complex, but the above is sufficient to show the issue. (My real model had Pads and Pockets after the LinearPattern.)

Select the Cylinder and export to STEP; the color shows up correctly. Select the Body and export to STEP; the color is missing. Delete the Cylinder, select the Body, and export to STEP; the color shows up correctly.

I can work around the issue by selecting the last feature of the problematic body (and, because I'm picky that way, manually editing the `.stp` to fix the part name), but in a complex model, that's a bit annoying. :slightly_smiling_face:

### Full version info

```shell
OS: Fedora Linux 38 (Thirty Eight) (KDE/plasmax11)
Word size of FreeCAD: 64-bit
Version: 0.20.2.Unknown
Build type: Unknown
Python 3.11.1, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.5, OCC 7.6.3
Locale: English/United States (en_US)
Installed mods: 
  * fasteners 0.4.65
  * kicadStepUpMod 10.19.4
```

### Subproject(s) affected?

None

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by luzpaz** [comment]
Please reproduce on latest stable (v0.21.1) and report back

**Comment by mwoehlke** [comment]
@luzpaz, I seem unable to fetch from GitHub at this time? In any case, it is going to be significant effort for me to try to build from source.

It takes *maybe* three minutes to test (less time than I've already spent trying to fetch the source), given the above test case. Did you try it? Were you unable to reproduce in v0.21.x?

**Comment by mwoehlke** [comment]
Anyway, neither 0.21.1 nor master are building for me. The error (same in both branches) is quite inscrutable:
```
In file included from freecad/src/CXX/Objects.hxx:42,
                 from freecad/src/Base/PyObjectBase.h:56,
                 from freecad/src/Base/BaseClass.cpp:31:
freecad/src/CXX/Python3/Objects.hxx:1336:21: error: expected unqualified-id before ‘const’
 1336 |         SeqBase<T>( const Object &ob )
      |                     ^~~~~
```

**Comment by luzpaz** [comment]
@Roy-043 do you mind testing this. I'm currently mobile and can't.

**Comment by mwoehlke** [comment]
Right, and it eventually dawns on me that `SeqBase<T>( const Object &ob )` looks dodgy. At least, GCC ≥ 11 doesn't like it with `--std=c++20`. I'll try to take another look now that I'm past that build failure...

**Comment by wwmayer** [comment]
There is an issue with gcc when trying to build as C++20. You have two choices to handle it:
* build with C++17 or
* use clang compiler

**Comment by wwmayer** [comment]
> Select the Cylinder and export to STEP; the color shows up correctly. Select the Body and export to STEP; the color is missing. Delete the Cylinder, select the Body, and export to STEP; the color shows up correctly.

I can 100% confirm your observations.

**Comment by mwoehlke** [comment]
> There is an issue with gcc when trying to build as C++20.

If by "issue" you mean GCC is obeying the standard. Filed as #10952.

> It has nothing to do with linear pattern. To me it looks like if the body has more than two objects the export has problems.

Could be; that just sounds like I didn't reduce my test case as much as was possible. :slightly_smiling_face:

Thanks for looking into this!

**Comment by luzpaz** [comment]
> It has nothing to do with linear pattern. If instead of a pattern you create a second cube and attach it to any of the faces of the first cube and then create a cylinder or third cube the STEP export loses the colors. To me it looks like if the body has more than two objects the export has problems.

Do we need to improve the export logic ?

**Comment by FEA-eng** [comment]
> Select the Body and export to STEP; the color is missing.

For me, it keeps the color. Is it possible that it got fixed?

<img width="857" height="403" alt="Image" src="https://github.com/user-attachments/assets/a845df0f-323f-4d13-ae10-9f531fe23e0f" />

```
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.1.0dev.14555 (Git shallow)
Build date: 2025/09/16 20:38:23
Build type: Release
Branch: main
Hash: bac49e465908071e7400a547eb813bc5017ae7dc
Python 3.11.13, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: Polish/Poland (pl_PL)
Navigation Style/Orbit Style/Rotation Mode: CAD/Rounded Arcball/Window center
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/90.0633/1
```

**Comment by luzpaz** [comment]
@mwoehlke ping.. please retest on 1.1RC1 https://github.com/FreeCAD/FreeCAD/releases/tag/1.1rc1

**Comment by ElementW** [comment]
Could reproduce in (all versions here are the Linux AppImages) 0.20, 0.21, 1.0.0, and 1.1.3, but not `weekly-2025.05.26`, the oldest weekly I have on hand at the moment, or any weeklies after that.

Clearly this existed at the 1.1.0 branch point but was fixed by something between it and the 2025.05.26 weekly.

**Comment by ipatch** [comment]
could not reproduce using the file linked above by @FEA-eng 

used a combination of legacy and non legacy step export options, and opening / importing the `.step` files the color stays intact.

```
OS: Fedora Linux Asahi Remix 44 (Forty Four) (tty)
Architecture: arm64
Version: 26.3.0dev.48649 (Git)
Build date: 2026/09/13 16:32:11
Build type: Release
Branch: main
Hash: e273f684f85ba2b42214fc6b831e5a106c0cec8a
Python 3.13.15, Qt 6.11.1, Coin 4.0.10 (bundled), Pivy 0.6.11 (bundled), Vtk 9.5.2, boost 1_92, Eigen3 5.0.1, PySide 6.11.1
shiboken 6.11.1, SMESH 7.7.1.0, xerces-c 3.3.0, Clipper2 , IfcOpenShell 0.8.4.post1, OCC 7.9.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
QPA/File dialog/Color dialog: xcb/non-native/non-native
Navigation Style/Orbit Style/Rotation Mode: Blender/Rounded Arcball/Window center
Logical DPI/Physical DPI/Pixel Ratio: 96/95.958/1
OpenGL version/vendor/renderer: 4.6 (Compatibility Profile) Mesa 26.2.1/Mesa/Apple M1 (G13G B1)
```

**Comment by luzpaz** [comment]
Please add full About info for builds you test against. It saves time and provides a lot of clarity for posterity.

**Comment by ElementW** [comment]
Well it's something among the many commits of [1.0.0...weekly-2025.04.21](https://github.com/FreeCAD/FreeCAD/compare/1.0.0...weekly-2025.04.21) (not d876e188082c314299f338799512965f3e25c03b because it's been backported to 1.0).

@luzpaz For versions, FC 1.1.3 is affected but not `weekly-2025.05.26`, yet:
* both use Qt 6.8.3
* both use Coin 4.0.3
* both use OCC 7.8.1

And I'm not going to drop 15+ full version dumps I've tested am I? (I mean, too late now, I'm not doing it after the fact)

**Comment by luzpaz** [comment]
Thanks. that's fine, IMHO. No, that wouldn't make sense and thank you for all your work.

**Comment by Roy-043** [comment]
Let's close then.
